### PR TITLE
[RFC] JHipster companion sidecar app for administrators to manage JHipster apps

### PR DESCRIPTION
**Pull request by PierreBesson**
@jhipster/developers Here is the RFC for my proposal.

### Summary

As of JHipster 6.5, the JHipster administrator's management and monitoring UI is spread out to various places (JHipster Registry, angular/react/vue.js pages) which causes code duplication and restraints their evolution. This RFC proposes to move this functionality to an external application that can be used in different development and production scenarios. In the following, this application will be referred to as the **JHipster companion**.

---

From here, this is what we have to do, see steps 6 and 7 of the RFC process in CONTRIBUTING.md.
> - Build consensus and integrate feedback from the reviewers
> - The Pull Request is either accepted (merged), rejected (closed) or postponed (given an "on hold" status)

I have temporarily named the project **JHipster Companion**, I will be launching a poll on Twitter about the name of the companion app.

However, we also have to decide the front-end framework that will be used to develop the app. Please use emojis to vote for Angular (:heart:), React (:rocket:), Vue.js (:tada:).

### CODE REVIEW

**Comment by deepu105** [comment]
can we skip builds for these

**Comment by pascalgrimaud** [comment]
I don't know how to do it... so if someone find the syntax for [skip ci]...

File: rfcs/2-jhipster-rfc-jhipster-companion.md [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+- [Rationale and alternatives](#rationale-and-alternatives)
+- [Prior art](#prior-art)
+- [Unresolved questions](#unresolved-questions)
+- [Future possibilities](#future-possibilities)
+
+# Summary
+
+[summary]: #summary
+
+As of JHipster 6.5, the JHipster administrator's management and monitoring UI is spread out to various places (JHipster Registry, angular/react/vue.js pages) which causes code duplication and restraints their evolution. This RFC proposes to move this functionality to an external application that can be used in different development and production scenarios. In the following, this application will be referred to as the **JHipster companion**.
```
Comment (MathieuAA) [inline rfcs/2-jhipster-rfc-jhipster-companion.md]: restraints => restrains
maybe it's too "heavy" a meaning for what you meant, what about "slows down"? Strong enough?

File: rfcs/2-jhipster-rfc-jhipster-companion.md [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+In recent years the technology landscape has changed significantly for Java and non Java based microservices but JHipster has somehow failed to adapt because we invested too much in maintaining our Eureka based solution: the JHipster registry. Although this solution is great and very productive for developers, it is not a very popular choice nowadays and our support for microservices without Eureka is lacking. Moreover, if we want to support more languages and technologies (.NET, Node.JS, Micronaut, Quarkus) and integrate them in the JHipster platform, non portable technologies such as Spring Cloud and Eureka cannot be made mandatory.
+
+In the meantime, our experience with the JHipster Registry has showed that in a microservice architecture, it is very valuable to have an overview of your running applications instances either in development or production. Currently this overview is provided as part of the application itself (for JHipster monoliths) or the JHipster Registry (for JHipster microservices). The goal of this RFC is to describe a solution that provide this overview in a new external application which can be used to cover all current use cases and more.
```
Comment (MathieuAA) [inline rfcs/2-jhipster-rfc-jhipster-companion.md]: that provide => that provides

File: rfcs/2-jhipster-rfc-jhipster-companion.md [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+- The application will be packaged as a docker image and a jar file which will be published on Docker Hub and Maven Central
+- After the initial **JHipster companion** 1.0 release, it will be expected to keep backward compatibility with JHipster applications generated with JHipster version 7.0 or higher.
+- The **JHipster companion** will be an optional part of JHipster so that users will not have to use it. However, we think that it will provide sufficient improvements to the development and operational experience to convince users to adopt it.
+
+
+# Drawbacks
+
+[drawbacks]: #drawbacks
+
+- As JHipster is first and foremost a code generator, it could be argued that it is not the goal of the project to release non code generator products. Two date, several external products have been released as part of the JHipster organisation: the JHipster Registry and JHipster Console and their release cadence has been low compared to the main generator.
```
Comment (MathieuAA) [inline rfcs/2-jhipster-rfc-jhipster-companion.md]: two date => perhaps you meant "to date"?

File: rfcs/2-jhipster-rfc-jhipster-companion.md [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+- Impose the overhead of running another service in development and production
+
+# Rationale and alternatives
+
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+Compared to today's situation, the proposed solution would have the following advantages :
```
Comment (MathieuAA) [inline rfcs/2-jhipster-rfc-jhipster-companion.md]: no space between "advantages" and the ":" char

File: rfcs/2-jhipster-rfc-jhipster-companion.md [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+Implementation details will be left to the discretion of the **JHipster companion** developers but the following guidelines will need to be followed:
+
+- A separate  **JHipster companion** GitHub repository will host the application and allow collaboration
+- The **JHipster companion** application will be based on a reactive Spring Boot application using Spring Cloud Gateway as embedded proxy technology
+- The application will include Spring Cloud dependencies to allow it to plug into Eureka, Consul and Kubernetes service discovery mechanism
+- The Management UI will be developped using Angular/React/Vue.js and can reuse some the existing code that currently lives in the "Admin pages" of a JHipster application
```
Comment (SudharakaP) [inline rfcs/2-jhipster-rfc-jhipster-companion.md]: developped --> developed

File: rfcs/2-jhipster-rfc-jhipster-companion.md [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+# Prior art
+
+[prior-art]: #prior-art
+
+Some existing similar or related solutions :
+
+- [JHipster Registry](https://www.jhipster.tech/jhipster-registry/)
+- [Spring Boot Admin](https://codecentric.github.io/spring-boot-admin/current/)
+- [Trampoline](https://ernestort.github.io/Trampoline/)
+- [Pivotal Cloud Fondry Apps Manager](https://docs.run.pivotal.io/console/manage-apps.html)
```
Comment (SudharakaP) [inline rfcs/2-jhipster-rfc-jhipster-companion.md]: Cloud Fondry --> Cloud Foundry

File: rfcs/2-jhipster-rfc-jhipster-companion.md [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+- How do we help transition legacy JHipster applications (version lower than 7.0) to be supported by the **JHipster companion** ?
+
+# Future possibilities
+
+[future-possibilities]: #future-possibilities
+
+Possible future evolutions:
+
+- Provide a plugin mechanism to let organizations customize the **JHipster companion** with custom features without forking
+- Seemlessly integrate with observablility tools such as ELK, Grafana and Zipkin
```
Comment (SudharakaP) [inline rfcs/2-jhipster-rfc-jhipster-companion.md]: observablility --> observability

Comment (SudharakaP) [inline rfcs/2-jhipster-rfc-jhipster-companion.md]: Seemlessly --> Seamlessly

File: rfcs/2-jhipster-rfc-jhipster-companion.md (hunk 3 above)
Comment (SudharakaP) [inline rfcs/2-jhipster-rfc-jhipster-companion.md]: Two date, --> To date,

File: rfcs/2-jhipster-rfc-jhipster-companion.md [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+# Rationale and alternatives
+
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+Compared to today's situation, the proposed solution would have the following advantages :
+
+- Reduce the generated code as code related to  management screens will no longer be generated
+- Faster build times for the front end resulting from less front-end code to compile
+- Stop developing the same management UI with 3 different frontend framework (Angular, React and Vue.js)
+- Improve security as management endpoint will no longer be exposed to the outside but only internally on a separate port. Also the **JHipster companion** can have proper authentication (jwt,oauth) rather than the basic authentication scheme that is present on the JHipster registry (which is the only scheme supported by the eureka server).
```
Comment (SudharakaP) [inline rfcs/2-jhipster-rfc-jhipster-companion.md]: Space missing: (jwt,oauth) --> (jwt, oauth)

File: rfcs/2-jhipster-rfc-jhipster-companion.md (hunk 3 above)
Comment (SudharakaP) [inline rfcs/2-jhipster-rfc-jhipster-companion.md]: organisation --> organization or change the organization to organisation in the **Possible future evolutions** section for consistency.

File: rfcs/2-jhipster-rfc-jhipster-companion.md (hunk 8 above)
Comment (SudharakaP) [reply rfcs/2-jhipster-rfc-jhipster-companion.md]: Also, probably, (JWT, OAuth) is better.

File: rfcs/2-jhipster-rfc-jhipster-companion.md [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+
+## Usage in development
+
+In every JHipster app, docker-compose files are present in the `src/main/docker` folder to launch associated services such as databases, message brokers and registries using docker. Among them, a **jhipster-companion.yml** will let the users starts the **JHipster companion** which will then be available at [http://localhost:1337](http://localhost:1337). By default the docker-compose file will be generated in such a way that the companion will be able to connect to the locally running JHipster application out of the box. Another possible way to run the **JHipster companion** in development would be to develop a simple CLI that allows to start the **JHipster companion** server quickly. Such a CLI tool could be embedded into the JHipster package distributed on NPM.
+
+
+## Usage in production
+
+Once deployed on a production server, the **JHipster companion** can be configured to discover JHipster application instances automatically by plugging into service discovery backends (eureka, consul, kubernetes) or using a static list of application URLs provided in the configuration.
```
Comment (SudharakaP) [inline rfcs/2-jhipster-rfc-jhipster-companion.md]: (eureka, consul, kubernetes) -> (Eureka, Consul, Kubernetes) for consistency.

File: rfcs/2-jhipster-rfc-jhipster-companion.md [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+- As JHipster is first and foremost a code generator, it could be argued that it is not the goal of the project to release non code generator products. Two date, several external products have been released as part of the JHipster organisation: the JHipster Registry and JHipster Console and their release cadence has been low compared to the main generator.
+- Impose the overhead of running another service in development and production
+
+# Rationale and alternatives
+
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+Compared to today's situation, the proposed solution would have the following advantages :
+
+- Reduce the generated code as code related to  management screens will no longer be generated
```
Comment (SudharakaP) [inline rfcs/2-jhipster-rfc-jhipster-companion.md]: Getting nit picky; 😸 

to  management --> to management (remove additional space)

**Comment by gmarziou** [comment]
This is interesting, I have few questions:

- Will the registry be discontinued? What about the Spring Cloud Config server it embeds?
- If it's an internal app, is it mandatory for the companion to embed a gateway?

**Comment by atomfrede** [comment]
One question that came into my mind yesterday, how does it differ from or relate to [spring boot admin](https://github.com/codecentric/spring-boot-admin) and what will be the unique selling point of our companion app (except of course eventually being not tight to spring boot anymore)?

**Comment by PierreBesson** [comment]
Thanks @SudharakaP @MathieuAA for the review.

@gmarziou 
>Will the registry be discontinued? What about the Spring Cloud Config server it embeds?

For now, I don't think we will discontinue the registry, there is no other "packaged solution" that provides an Eureka server. However, we will be able to remove the management pages from the Registry UI. As for centralized configuration, there is no plan to support it in the JHipster Companion for now but we can think about it in the future.

>If it's an internal app, is it mandatory for the companion to embed a gateway?

If it doesn't embed a gateway, I don't see how you would access apps management endpoints securely and without using CORS.

@atomfrede 
> how does it differ from or relate to spring boot admin ?

It does provide the same set of features but will be more integrated with JHipster and have a more polished and complete experience (for example it will also include API docs). It must also work out of the box with no setup from the user. The UX needs to be simple and allow to easily go from an overview of all application instances to specific instance info. It's also very important to implement the UI router correctly so that management screens can have shareable URLs.

Ultimately my proposal is only to reimplement most of what is already available in the JHipster Registry but covering all use cases and in a tech agnostic way.

**Comment by gmarziou** [comment]
Calling it sidecar makes me wonder whether there will be only one for multiples apps like the registry or one per app.

**Comment by PierreBesson** [comment]
@gmarziou Yes that's why I don't like the term sidecar too much because you only need 1 companion instance for multiples microservices not one for every running instance.

**Comment by deepu105** [comment]
Hmmm, I liked the name sidecar, but after reading the comment from Gael, I
think it makes sense not to call it sidecar, so ya we need to come up with
a better name.
Companion is too generic and not that hip IMO, we also need to think of the
marketing value, unfortunately, so let's not worry too much for the name
now, let's assign a code name for now, once we have solid work we can think
of a name.

Thanks & Regards,
Deepu

**Comment by gmarziou** [comment]
Any outcome from the twitter pol about names?

**Comment by PierreBesson** [comment]
In the twitter poll (around 50 respondents), the most popular choice are: JHipster Control Center and JHipster Admin. The problem with JHipster Admin is that people might thing that it's too similar to Spring Boot Admin.

**Comment by deepu105** [comment]
JHipster Control Center doen't sound bad

Thanks & Regards,
Deepu

**Comment by jdubois** [comment]
I have spent some time working on Dapr https://github.com/jhipster/generator-jhipster/issues/10958 recently (that's part of my day job, yay!).
It might be a solution for our microservices needs, so we remove the difference between microservices and monoliths, and so we become language and framework agnostic. Then it would tie us very closely to this solution, which is something I don't like too much, as it's not widespread or standard.

**Comment by PierreBesson** [comment]
I agree that being platform agnostic is the right direction. But then why choose Microsoft Dapr instead of Google Knative ?
For me those projects are too young, let's focus on improving our support for mature tools like native Kubernetes and Helm. They can be supported through additional sub-generators.

**Comment by PierreBesson** [comment]
Okay so here are the decisions, one week after launching the votes :
- The application will be named the **JHipster Control Center**
- The frontend will use the **Vue.js** framework.
- The backend will use **Java** rather than Kotlin, to make sure that a maximum number of people can contribute

Screenshot from 2019-12-19 11-50-45

I'm amending the RFC to add those elements into account.

**Pull request merged**